File: 0819. Most Common Word.java
class Solution {
    public String mostCommonWord(String para, String[] banned) {
        HashMap<String,Integer> frq=new HashMap<>();
        para+='.';  //in case in doesn't end with punctuation, this ensures last word will be processed
        for(String b:banned) frq.put(b,-1); //mark as banned
        int st=0;
        int max=0;
        String res="";
        for(int i=0;i!=para.length();i++){
            char c=para.charAt(i);
            if(c==' '||c=='!'||c=='?'||c==','||c==';'||c=='.'||c=='\''){    //end of word detected
                String word=para.substring(st,i).toLowerCase();
                st=i+1;
                if(word.length()==0||frq.get(word)!=null&&frq.get(word)==-1) continue;   
                //no word (ex. ", " would give an empty string) or banned word, don't process
                // System.out.println(word);
                frq.put(word,frq.getOrDefault(word,0)+1);   //update freq of this word
                int ct=frq.get(word);
                if(ct>max){ //if most frequent valid ord
                    max=ct;
                    res=word;
                }
            }
        }
        return res;
    }
}

File: 0819. Most Common Word.cpp
class Solution {
    public:
        string lower(string str) {
            string res;
            for (int i = 0; i != str.length(); i++) res += tolower(str[i]);
            return res;
        }
    string mostCommonWord(string para, vector < string > & banned) {
        map < string, int > frq;
        for (string b: banned) frq[b] = -1;
        para += '.';
        int st = 0;
        int max = 0;
        string res;
        for (int i = 0; i != para.length(); i++) {
            char c = para[i];
            // cout<<i<<" "<<st<<endl;
            if (c == ' ' || c == '!' || c == '?' || c == ',' || c == ';' || c == '.' || c == '\'') {
                string word = lower(para.substr(st, i - st > 0 ? i - st : 0));
                // cout<<c<<word<<endl;
                st = i + 1;
                if (word.length() == 0 || frq[word] == -1) continue;
                frq[word]++;
                int ct = frq[word];
                if (ct > max) {
                    max = ct;
                    res = word;
                }
            }
        }
        return res;
    }
};
